src/diffusers/pipelines/blip_diffusion/pipeline_blip_diffusion.py

```diff
@@ -0,0 +1,431 @@
+        self.final_offload_hook = hook
+
+    @property
+    # Copied from diffusers.pipelines.stable_diffusion.pipeline_stable_diffusion.StableDiffusionPipeline._execution_device
+    def _execution_device(self):
+        r"""
+        Returns the device on which the pipeline's models will be executed. After calling
+        `pipeline.enable_sequential_cpu_offload()` the execution device can only be inferred from Accelerate's module
+        hooks.
+        """
+        if not hasattr(self.unet, "_hf_hook"):
+            return self.device
+        for module in self.unet.modules():
+            if (
+                hasattr(module, "_hf_hook")
+                and hasattr(module._hf_hook, "execution_device")
+                and module._hf_hook.execution_device is not None
+            ):
+                return torch.device(module._hf_hook.execution_device)
+        return self.device
+
+
+    def _build_prompt(self, prompts, tgt_subjects, prompt_strength=1.0, prompt_reps=20):
+        rv = []
+        for prompt, tgt_subject in zip(prompts, tgt_subjects):
+            prompt = f"a {tgt_subject} {prompt.strip()}"
+            # a trick to amplify the prompt
+            rv.append(", ".join([prompt] * int(prompt_strength * prompt_reps)))
+
+        return rv
+
+    def _predict_noise(
+        self,
+        t,
+        latent_model_input,
+        text_embeddings,
+        width=512,
+        height=512,
+        cond_image=None,
+    ):
+ 
+        down_block_res_samples, mid_block_res_sample = None, None
+
+        noise_pred = self.unet(
+            latent_model_input,
+            timestep=t,
+            encoder_hidden_states=text_embeddings,
+            down_block_additional_residuals=down_block_res_samples,
+            mid_block_additional_residual=mid_block_res_sample,
+        )["sample"]
+
+        return noise_pred
+
+    def _init_latent(self, latent, height, width, generator, batch_size):
+        if latent is None:
+            latent = torch.randn(
+                (1, self.unet.in_channels, height // 8, width // 8),
+                generator=generator,
+                device=generator.device,
+            )
+        latent = latent.expand(
+            batch_size,
+            self.unet.in_channels,
+            height // 8,
+            width // 8,
+        )
+        return latent.to(self.device)
+
+    def _forward_prompt_embeddings(self, input_image, src_subject, prompt):
+        # 1. extract BLIP query features and proj to text space -> (bs, 32, 768)
+        query_embeds = self.forward_ctx_embeddings(input_image, src_subject)
+
+        # 2. embeddings for prompt, with query_embeds as context
+        tokenized_prompt = self._tokenize_text(prompt).to(self.device)
+        text_embeddings = self.text_encoder(
+            input_ids=tokenized_prompt.input_ids,
+            ctx_embeddings=query_embeds,
+            ctx_begin_pos=[self._CTX_BEGIN_POS],
+        )[0]
+
+        return text_embeddings
```

Comment:
I think there are two things in here. WDYT about having an `encode_prompt()` method and  a separate `get_query_embeddings()` method? 

This way, we would promote a better separation of concern and composability. 

This will also allow the users to explicitly pass the `query_embeds` for whatever reason (memory savings, etc.). 

Or we, can directly make use of `query_embeds = self.forward_ctx_embeddings(input_image, src_subject)` and propagate `query_embeds` to `encode_prompt()` subsequently. 

Does this make sense?

Reply:
I think it makes sense to have a separate `get_query_embeddings` and `encode_prompt`, to enable getting the query embeds if necessary, caching them and then passing to `encode_prompt`, if something like that is required.